sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala

```diff
@@ -58,8 +58,21 @@ object Row {
    * Merge multiple rows into a single row, one after another.
    */
   def merge(rows: Row*): Row = {
-    // TODO: Improve the performance of this if used in performance critical part.
-    new GenericRow(rows.flatMap(_.toSeq).toArray)
+    val size = rows.size
+    var number = 0
+    for (i <- 0 until size) {
+      number = number + rows(i).size
+    }
+    val container = Array.ofDim[Any](number)
+    var n = 0
+    for (i <- 0 until size) {
```

Comment:
if you want to improve this more, you can try with a while loop here rather than a for loop, without creating the extra 0 until size.

Reply:
Only primitively use size, subSize, and number information and control the container will improve the performance more.
up to 
call 100000000 time Row.merge(row1) need 18064 millisecond
call 100000000 time Row.merge(rows:_*) need 25651 millisecond